src/objective_types.jl

```diff
@@ -0,0 +1,271 @@
+            ReverseDiff.gradient!(gr_res, f, x, gcfg)
+            DiffBase.value(gr_res)
+        end
+        hcfg = ReverseDiff.HessianConfig(x_seed)
+        h! = (x, out) -> ReverseDiff.hessian!(out, f, x, hcfg)
+    else
+        error("The autodiff value $(autodiff) is not supported. Use :finite, :forward or :reverse.")
+    end
+    g = similar(x_seed)
+    H = Array{T}(n_x, n_x)
+    g!(x_seed, g)
+    h!(x_seed, H)
+    return TwiceDifferentiable(f, g!, fg!, h!, f(x_seed),
+                                       g, H, copy(x_seed),
+                                       copy(x_seed), copy(x_seed), f_calls, g_calls, h_calls)
+end
+
+
+function TwiceDifferentiable{T}(f, g!, x_seed::Array{T}; autodiff = :finite)
+    n_x = length(x_seed)
+    f_calls = [1]
+    function fg!(x, storage)
+        g!(x, storage)
+        return f(x)
+    end
+    if autodiff == :finite
+        function h!(x, storage)
+            Calculus.finite_difference_hessian!(f, x, storage)
+            return
+        end
+    elseif autodiff == :forward
+        hcfg = ForwardDiff.HessianConfig(similar(x_seed))
+        h! = (x, out) -> ForwardDiff.hessian!(out, f, x, hcfg)
+    elseif autodiff == :reverse
+        hcfg = ReverseDiff.HessianConfig(x_seed)
+        h! = (x, out) -> ReverseDiff.hessian!(out, f, x, hcfg)
+    else
+        error("The autodiff value $(autodiff) is not supported. Use :finite, :forward or :reverse.")
+    end
+    g = similar(x_seed)
+    H = Array{T}(n_x, n_x)
+    g!(x_seed, g)
+    h!(x_seed, H)
+    return TwiceDifferentiable(f, g!, fg!, h!, f(x_seed),
+                                       g, H, copy(x_seed),
+                                       copy(x_seed), copy(x_seed), f_calls, [1], [1])
+end
+
+function TwiceDifferentiable(d::OnceDifferentiable; autodiff = :finite)
+    n_x = length(d.last_x_f)
+    T = eltype(d.last_x_f)
+    if autodiff == :finite
+        function h!(x::Vector, storage::Matrix)
+            Calculus.finite_difference_hessian!(d.f, x, storage)
+            return
+        end
+    elseif autodiff == :forward
+        hcfg = ForwardDiff.HessianConfig(similar(gradient(d)))
+        h! = (x, out) -> ForwardDiff.hessian!(out, d.f, x, hcfg)
+    end
+    H = Array{T}(n_x, n_x)
+    h!(d.last_x_g, H)
+    return TwiceDifferentiable(d.f, d.g!, d.fg!, h!, d.f_x,
+                                       gradient(d), H, d.last_x_f,
+                                       d.last_x_g, copy(d.last_x_g), d.f_calls, d.g_calls, [1])
+end
+
+function _unchecked_value!(obj, x)
+    obj.f_calls .+= 1
+    copy!(obj.last_x_f, x)
+    obj.f_x = obj.f(x)
+end
+function value(obj, x)
+    if x != obj.last_x_f
+        obj.f_calls += 1
+        return obj.f(x)
+    end
+    obj.f_x
+end
+function value!(obj, x)
+    if x != obj.last_x_f
+        _unchecked_value!(obj, x)
+    end
+    obj.f_x
+end
+
+
+function _unchecked_grad!(obj, x)
+    obj.g_calls .+= 1
+    copy!(obj.last_x_g, x)
+    obj.g!(x)
```

Comment:
Yes, you're right, thanks. I actually did notice this myself, but apparently forgot to change it. They aren't used (yet) as all gradient updates happen through `fg!`.